library/common.h

```diff
@@ -114,6 +114,18 @@ extern void (*mbedtls_test_hook_test_fail)(const char *test, int line, const cha
  */
 #define MBEDTLS_ALLOW_PRIVATE_ACCESS
 
+/**
+ * \brief       Securely zeroize a buffer then free it.
+ *
+ *              Exactly the same as consecutive calls to
+ *              \c mbedtls_platform_zeroize() and \c mbedtls_free(), but has a
+ *              code size savings, and potential for optimisation in the future.
+ *
+ * \param buf   Buffer to be zeroized then freed. May be \c NULL.
```

Comment:
Rather than “may be NULL”, I would say: does not have to be a valid pointer if `len` is 0. I would not expect this function to have defined behavior if `buf == NULL` but `len > 0`.

Related: https://github.com/Mbed-TLS/mbedtls/issues/6633

Reply:
Do you mean "defined" or "specified"?

I really don't want this function to ever have undefined behaviour.

Since `free(NULL)` is defined as a no-op, I would actually like `mbedtls_zeroize_and_free(NULL, x)` to be a no-op for any value of `x`.

The benefit: tests for `!= NULL` can be removed before calling this function.

The down side: silently ignoring a `NULL` value, rather than crashing. But given that this operation is typically at the end of a function, I don't think we lose much there, and we gain by not having any chance of undefined behaviour.

Would you accept a change to `If NULL, no action is taken`?

Reply:
> I really don't want this function to ever have undefined behaviour.

That's impossible: it has a pointer argument, and you want to do something useful with that pointer.

> I would actually like mbedtls_zeroize_and_free(NULL, x) to be a no-op for any value of x

`mbedtls_zeroize_and_free(NULL, x)` for `x > 0` is likely to be a memory management error and I'd want it to cause a null pointer dereference exception at runtime on platforms where that's a thing.

> we gain by not having any chance of undefined behaviour.

I don't see how we gain by silently accepting a null pointer with a nonzero size.

Reply:
Actually, no, I disagree.

It's common in C to use `p == NULL` to mean "there is no buffer (yet)". At which point "length" is meaningless (and the variable/field to hold it may not have been initialised).

For example

```
char *p = NULL;
char *q = NULL;
int m, n;       /* lengths of p and q */
int ret = -1;

m = space_needed_for_p();
if (m == 0) goto exit;

p = malloc(m);
if (p == NULL) goto exit;

n = space_needed_for_q();
q = malloc(n);
if (q == NULL) goto exit;

:
:  /* do stuff with (p, m) and (q, l) */
:

ret = 0;

exit:
free(p);
free(q);
return ret;
```

This is why `free(NULL)` was defined in the standards as a no-op - it didn't used to be (way back in the midst of time when I started with C) and it would be common to see `if (p != NULL) free(p)` all over the place. Thankfully we don't have to do that now.

This implementation of `mbedtls_zeroize_and_free()` is intended to an extension of `free()` rather than an extension of `bzero_explicit()`, so it should silently ignore buffers with `NULL`, rather than require they have length `0`.

And is test is definitely needed, since `memset(NULL, 0, 0)` is undefined behaviour, even if we require `NULL` to always have length `0` we can't call `memset()`.

So in my view much better to test for `buf != NULL` than `len != 0`.

Now, for parameters to a function, where the caller specifies (buffer, length) it makes sense to require that if `buffer == NULL` then `length == 0`.

But within a function - and this is an internal library function - that's not necessarily the case.

We definitely use the "buffer == NULL => no buffer" within our code

```
        if (operation->ctx.pbkdf2.salt != NULL) {
            mbedtls_platform_zeroize(operation->ctx.pbkdf2.salt,
            mbedtls_free(operation->ctx.pbkdf2.salt);
        }
```

and it's even done in PSA:

```
void psa_free_persistent_key_data(uint8_t *key_data, size_t key_data_length)
{
    if (key_data != NULL) {
        mbedtls_platform_zeroize(key_data, key_data_length);
    }
    mbedtls_free(key_data);
}
```

and looking at this PR there are several places where I can remove `!= NULL` checks before calling `mbedtls_zeroize_and_free()`, which would definitely be code savings

Reply:
@tom-cosgrove-arm In all the code snippet you cite as examples of `buffer == NULL` conditionals, there is a precondition that `buffer` is a pointer to a buffer of `length` bytes, hence `buffer == NULL` can only happen when `length == 0`. `zeroize` is a no-op when `length == 0` but [undefined behavior when the pointer is null](https://github.com/Mbed-TLS/mbedtls/issues/6633), so the test to call it can be either `length != 0` or `buffer != NULL`. However, when the pointer is non-null, `free(buffer)` is not a no-op (it may be an implementation where `malloc(0)` is equivalent to `malloc(1)`) so the test to call `zeroize` and `free` has to be on the pointer.

You propose to make `mbedtls_platform_zeroize_and_free` a no-op when `buffer == NULL` and `length != 0`, i.e. in a condition that violates a habitual invariant. I would rather signal that the invariant has been violated, rather than allow what is probably a coding error.

> It's common in C to use p == NULL to mean "there is no buffer (yet)". At which point "length" is meaningless (and the variable/field to hold it may not have been initialised).

It isn't common in our code base. I'm not saying it never happens, but we usually treat (buffer, length) as a pair that's modified together, with the invariant that either `buffer` points to `length` bytes of memory or `buffer == NULL && length == 0`. Just because you _can_ make `NULL` mean “not there yet” [doesn't mean it's a good idea](https://en.wikipedia.org/wiki/Null_pointer#History).

Reply:
So @gilles-peskine-arm are you saying that at all the places in our code where we have (for example)

```
        if (operation->ctx.pbkdf2.salt != NULL) {
            mbedtls_platform_zeroize(operation->ctx.pbkdf2.salt,
                                     operation->ctx.pbkdf2.salt_length);
            mbedtls_free(operation->ctx.pbkdf2.salt);
        }
```

if we would guarantee that `mbedtls_platform_zeroize(NULL, 0)` was a no-op but `mbedtls_platform_zeroize(NULL, 1)` could crash, we can just write

```
        mbedtls_platform_zeroize(operation->ctx.pbkdf2.salt,
                                 operation->ctx.pbkdf2.salt_length);
        mbedtls_free(operation->ctx.pbkdf2.salt);
```

and the only difference would be performance? i.e. if `salt == NULL` => `salt_length == 0`?